mm2src/lp_ordermatch/simple_market_maker.rs

```diff
@@ -0,0 +1,734 @@
+    CannotStartFromStopping,
+    #[display(fmt = "Internal error: {}", _0)]
+    InternalError(String),
+}
+
+#[derive(Debug)]
+pub enum PriceServiceRequestError {
+    HttpProcessError(String),
+    ParsingAnswerError(String),
+}
+
+impl From<serde_json::Error> for PriceServiceRequestError {
+    fn from(error: serde_json::Error) -> Self { PriceServiceRequestError::ParsingAnswerError(error.to_string()) }
+}
+
+impl From<std::string::String> for PriceServiceRequestError {
+    fn from(error: String) -> Self { PriceServiceRequestError::HttpProcessError(error) }
+}
+
+impl From<std::str::Utf8Error> for PriceServiceRequestError {
+    fn from(error: Utf8Error) -> Self { PriceServiceRequestError::HttpProcessError(error.to_string()) }
+}
+
+impl HttpStatusCode for StartSimpleMakerBotError {
+    fn status_code(&self) -> StatusCode {
+        match self {
+            StartSimpleMakerBotError::AlreadyStarted
+            | StartSimpleMakerBotError::InvalidBotConfiguration
+            | StartSimpleMakerBotError::CannotStartFromStopping => StatusCode::BAD_REQUEST,
+            StartSimpleMakerBotError::Transport(_) | StartSimpleMakerBotError::InternalError(_) => {
+                StatusCode::INTERNAL_SERVER_ERROR
+            },
+        }
+    }
+}
+
+impl HttpStatusCode for StopSimpleMakerBotError {
+    fn status_code(&self) -> StatusCode {
+        match self {
+            // maybe bad request is not adapted for the first errors.
+            StopSimpleMakerBotError::AlreadyStopped | StopSimpleMakerBotError::AlreadyStopping => {
+                StatusCode::BAD_REQUEST
+            },
+            StopSimpleMakerBotError::Transport(_) | StopSimpleMakerBotError::InternalError(_) => {
+                StatusCode::INTERNAL_SERVER_ERROR
+            },
+        }
+    }
+}
+
+#[derive(Clone)]
+struct TradingPair {
+    base: String,
+    rel: String,
+}
+
+impl TradingPair {
+    pub fn new(base: String, rel: String) -> TradingPair { TradingPair { base, rel } }
+
+    pub fn as_combination(&self) -> String { self.base.clone() + "/" + self.rel.clone().as_str() }
+}
+
+pub async fn tear_down_bot(ctx: MmArc) {
+    let simple_market_maker_bot_ctx = TradingBotContext::from_ctx(&ctx).unwrap();
+    {
+        let mut trading_bot_cfg = simple_market_maker_bot_ctx.trading_bot_cfg.lock().await;
+        trading_bot_cfg.clear();
+    }
+    cancel_pending_orders(&ctx).await;
+}
+
+async fn vwap_calculation(
+    kind: VwapCalculationSide,
+    swaps_answer: MyRecentSwapsResponse,
+    nb_valid_trades: &mut usize,
+    cfg: &SimpleCoinMarketMakerCfg,
+    calculated_price: MmNumber,
+) -> MmNumber {
+    let mut average_trading_price = calculated_price.clone();
+    let mut total_sum_price_volume = MmNumber::default();
+    let mut total_volume = MmNumber::default();
+    for swap in swaps_answer.swaps.iter() {
+        if !swap.is_finished_and_success() {
+            *nb_valid_trades -= 1;
```

Comment:
Are we protected from underflow here?

Reply:
> Are we protected from underflow here?

No because nb_valid_trades is equal to base_swaps.len() + rel_swaps.len() and we iterate over them in ` for swap in swaps_answer.swaps.iter() {`

If there are no successful swaps in the trade history, we will return the calculated price just after iteration.

Reply:
This looks like a dangerous side effect, which can be avoided with the following approach:
1. Remove `nb_valid_trades`.
2. Change `swaps_answer: MyRecentSwapsResponse` to `swaps_with_infos: impl IntoIterator<(SavedSwap, MySwapInfo)>`.
3. Preprocess the swaps *before* calling `vwap_calculation`, the code
```rust
    let base_swaps_with_infos = base_swaps.swaps.into_iter().filter_map(|swap| {
        if !swap.is_finished_and_success() {
            return None;
        }

        swap.get_my_info().map(|info| (swap, info))
    });
```

Reply:
I've changed my approach drastically; I realized that I don't need to know the number of valid trades based on the answer size; I can start from zero and increase each time I have a good swap. So I pushed the new approach - let me know if it's better now.

Reply:
It's a bit different from what I proposed, but anyway much better than it was before. Thanks! Resolved :slightly_smiling_face: